File: bridgePattern/abstraction.cpp
// ============= Refined Abstraction =============#include "common.h"
#include "implementation.h"
#include "./abstraction.h"
#include <iostream>

class ImplementationFactory
{
public:
	static WindowImp *CreateImplementation()
	{
		return new PMWindowImp();
	}

private:
	ImplementationFactory(){};
};

// ============= Abstraction =============
Window::Window()
{
	_imp = ImplementationFactory::CreateImplementation();
}

void Window::DrawRect()
{
	_imp->DeviceRect();
}

// ============= Refined Abstraction =============
void ApplicationWindow::DrawContents()
{
	std::cout << "Draw contents for ApplicationWindow" << std::endl;
}

// ============= Refined Abstraction =============
void IconWindow::DrawContents()
{
	std::cout << "Draw contents for IconWindow" << std::endl;
}

File: bridgePattern/csharp/Abstraction.cs
static class ImplementationFactory
{
	public static WindowImp CreateImplementation()
	{
		return new PMWindowImp();
	}
}

// ============= Abstraction =============
abstract class Window
{
	protected WindowImp _imp;

	public Window()
	{
		// Use factory method
		_imp = ImplementationFactory.CreateImplementation();
	}

	// requests handled by abstraction
	public abstract void DrawContents();
	// requests forwarded to implementation
	public void DrawRect()
	{
		_imp.DeviceRect();
	}


}

// ============= Refined Abstraction =============
class ApplicationWindow : Window
{
	public override void DrawContents()
	{
		System.Console.WriteLine("Draw contents for ApplicationWindow");
	}
}

class IconWindow : Window
{
	public override void DrawContents()
	{
		System.Console.WriteLine("Draw contents for IconWindow");
	}
}